Step 3: JSON

zghera opened the pull request:
Extends changes made in PR #10. Main change made was the integration of JSON messages into the servlet/JS data exchange rather than a single string via html. Modified the servlet to return a hardcoded list of strings using a JSON string. Then change JS code to fetch that JSON string, convert back to list, and display the next element in the list (wraps around) on each button press.

zghera committed:
Add JSON functionality.

Modify the servlet to return a hardcoded list in the form of a JSON string. Then change JS code to fetch that JSON string, convert back to list, and display the next element in the list (wraps around) on each button press.

zghera committed:
Proper url for main page reload.

beiebi0 commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -19,14 +19,33 @@
+  }
+
+  private String convertToJson(ArrayList<String> answers) {
+    Gson gson = new Gson();
```
This is cool!

zghera replied:
I agree!

beiebi0 commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -19,14 +19,33 @@
+/** Servlet that returns some example content. */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+  private final ArrayList<String> answers = new ArrayList<String>( 
```
```suggestion
  private static final List<String> ANSWERS =
      Arrays.asList("A stick!", "A wet log!", "(some) Frogs!");
```

zghera replied:
Done.

beiebi0 commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -19,14 +19,33 @@
+    response.getWriter().println(json);
+  }
+
+  private String convertToJson(ArrayList<String> answers) {
```
Loose it up to List<String> instead of ArrayList

zghera replied:
Done.

beiebi0 commented on portfolio/src/main/webapp/scripts/server-dev.js:
```diff
@@ -12,17 +12,16 @@
+  .then(response => response.json())
+  .then(answ_list => {
+    document.getElementById('answer-container').innerText = 
+                                          answ_list[(i++) % 3];
```
replace 3 with answ_list.size? or length. (sorry I forgot which is the right one)

zghera replied:
Done.

beiebi0 commented on portfolio/src/main/webapp/scripts/server-dev.js:
```diff
@@ -12,17 +12,16 @@
+  .then(response => response.json())
+  .then(answ_list => {
+    document.getElementById('answer-container').innerText = 
+                                          answ_list[(i++) % 3];
```
will this i++ go out of limit in any case?

zghera replied:
Hmmm. I did not think of that but I guess the index could theoretically exceed the max positive value and then overflow. I added a protection to reset to 0 once it reaches MAX_SAFE_INTEGER per an article I read.

beiebi0 commented on portfolio/src/main/webapp/scripts/server-dev.js:
```diff
@@ -12,17 +12,16 @@
-function addRandomGreeting() {
-  const greetings =
-      ['Hello world!', '¡Hola Mundo!', '你好，世界！', 'Bonjour le monde!'];
+var i = 0;
```
rename the variable here please to make it more meaningful. also probably add a comment for it if needed to explain what's that for.

zghera replied:
Done.

beiebi0 commented on walkthroughs/week-3-server/examples/random-quotes/src/main/webapp/script.js:
```diff
@@ -49,15 +49,33 @@ function addQuoteToDom(quote) {
     document.getElementById('quote-container').innerText = quote;
+  })
+  .catch(err => {
+    console.log("Error: " + err);
```
what will users see in this case? an error page? Or do you need a backup value for the innerText?

zghera replied:
Very good point. I added a backup value for the innerText. For the future in a real implementation, I think an error page would be best. Please let me know if you have a specific preference/suggestion.

zghera committed:
Reference local file rather than url

See commit f1a0efc.

zghera committed:
Switch to List for flexibility.

zghera committed:
Safegaurds for edge cases

In addition to better variable names, two main issues fixed: 1. Ensure there is no overflow of the answer index variable by resetting it to 0 if it gets above the MAX_SAFE_INTEGER. 2. Write a default message for the inner text in case an error is caught.

zghera committed:
Remove extra parenthesis bug.

zghera committed:
CSS style guideline changes from earlier PRs.

See 39c57c3.

beiebi0 commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -19,14 +19,30 @@
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import com.google.gson.Gson;
```
fix the order here

zghera replied:
I was not entirely sure what order I should have so I took my best guess: a1fda42.

beiebi0 replied:
https://google.github.io/styleguide/javaguide.html#s3.3-import-statements

zghera replied:
Okay that is what I thought it would be. I will make sure to skim over the style guides tomorrow for JS and Java.

beiebi0 commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -19,14 +19,30 @@
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+  public void doGet(HttpServletRequest request, HttpServletResponse response) 
+                                                          throws IOException {
```
Seems can fit one line here. No need to change.

zghera replied:
abf753c.

beiebi0 commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -19,14 +19,30 @@
+  public void doGet(HttpServletRequest request, HttpServletResponse response) 
+                                                          throws IOException {
+    // Convert the array of answers to JSON
+    String json_answers = convertToJson(ANSWERS);
```
lowerCamelCase for the variable name here

zghera replied:
abf753c.

beiebi0 commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -19,14 +19,30 @@
+
+  private String convertToJson(List<String> answers) {
+    Gson gson = new Gson();
+    String json = gson.toJson(answers);
```
Better naming here ? like answersInJson?

zghera replied:
abf753c.

beiebi0 commented on portfolio/src/main/webapp/scripts/server-dev.js:
```diff
@@ -0,0 +1,32 @@
+
+var answer_idx = 0;
+
+function getAnswer() {
+  fetch('/data')
+  .then(response => response.json())
+  .then(answers => {
+    document.getElementById('answer-container').innerText = 
+                              answers[(answer_idx++) % answers.length];
+    if(answer_idx >= Number.MAX_SAFE_INTEGER) {
+      answer_idx = 0;
+    }
+  })
+  .catch(err => {
+    console.log("Error: " + err);
+    document.getElementById('answer-container').innerText = 
+                                      "Something went wrong.";
+  });
+}
```
```suggestion
function getAnswer() {
  fetch('/data')
      .then(response => response.json())
      .then(answers => {
        document.getElementById('answer-container').innerText =
            answers[(answer_idx++) % answers.length];
        if (answer_idx >= Number.MAX_SAFE_INTEGER) {
          answer_idx = 0;
        }
      })
      .catch(err => {
        console.log('Error: ' + err);
        document.getElementById('answer-container').innerText =
            'Something went wrong.';
      });
}
```

beiebi0 commented on walkthroughs/week-3-server/examples/random-quotes/src/main/webapp/script.js:
```diff
@@ -49,15 +49,33 @@ function addQuoteToDom(quote) {
   quoteContainer.innerText = quote;
 }
 
+function getRandomQuoteUsingPromiseChain() {
+  fetch('/random-quote')
+  .then(function(response) { 
+    return response.text() 
+  })
+  .then(function(quote) {
+    document.getElementById('quote-container').innerText = quote;
+  })
+  .catch(function(err) {
+    console.log("Error: " + err);
+  });
+}
```
```suggestion
function getRandomQuoteUsingPromiseChain() {
  fetch('/random-quote')
      .then(function(response) {
        return response.text()
      })
      .then(function(quote) {
        document.getElementById('quote-container').innerText = quote;
      })
      .catch(function(err) {
        console.log('Error: ' + err);
      });
}
```

beiebi0 commented on walkthroughs/week-3-server/examples/random-quotes/src/main/webapp/script.js:
```diff
@@ -49,15 +49,33 @@ function addQuoteToDom(quote) {
+  })
+  .catch(err => {
+    console.log("Error: " + err);
   });
```
```suggestion
function getRandomQuoteUsingArrowFunctions() {
  fetch('/random-quote')
      .then(response => response.text())
      .then(quote => {
        document.getElementById('quote-container').innerText = quote;
      })
      .catch(err => {
        console.log('Error: ' + err);
      });
}
```

beiebi0 reviewed (commented):
For those JS files, I used an internal formatter to fix some of the JS formatting issues.

zghera commented:
> For those JS files, I used an internal formatter to fix some of the JS formatting issues.

In the future, is there something I could reference so I can avoid corrections on indentation?

zghera committed:
Alter import order to be alphabetical.

zghera committed:
Variable renaming and line char limit fix.

zghera committed:
Auto-formatter indentation fix on server_dev.js

Co-authored-by: beiebi0 <longjuntan@gmail.com>

zghera committed:
Auto-formatter indentation fix on example js script 1

Co-authored-by: beiebi0 <longjuntan@gmail.com>

zghera committed:
Auto-formatter indentation fix on example js script 2

Co-authored-by: beiebi0 <longjuntan@gmail.com>

zghera committed:
Merge branch master into step-3-json.

zghera committed:
Merge branch 'master' into step-3-json

beiebi0 commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -30,19 +30,18 @@
   }
 
-  private String convertToJson(List<String> answers) {
+  private String convertToJson(List<String> jsonAnswers) {
```
This one is probably better to have answers as you had before.

zghera replied:
Yeah that sounds good. Done.

beiebi0 reviewed (approved):
Left another comment but won't block your code merge. 
Fix that if you think that makes sense before you merge the code. Otherwise, good to go.

zghera committed:
Re-name convertToJson param.
